store/src/java/com/zimbra/cs/account/auth/AuthMechanism.java

```diff
@@ -235,30 +243,32 @@ public boolean isZimbraAuth() {
         public void doAuth(LdapProv prov, Domain domain, Account acct, String password,
                 Map<String, Object> authCtxt) throws ServiceException {
 
-            String encodedPassword = acct.getAttr(Provisioning.A_userPassword);
             if (AuthMechanism.doTwoFactorAuth(acct, password, authCtxt)) {
                 return;
             }
 
+            String encodedPassword = acct.getAttr(Provisioning.A_userPassword);
             if (encodedPassword == null) {
                 throw AuthFailedServiceException.AUTH_FAILED(acct.getName(),
                         namePassedIn(authCtxt), "missing "+Provisioning.A_userPassword);
             }
 
-            if (PasswordUtil.SSHA512.isSSHA512(encodedPassword)) {
-                if (PasswordUtil.SSHA512.verifySSHA512(encodedPassword, password)) {
-                    return; // good password, RETURN
-                }  else {
-                    throw AuthFailedServiceException.AUTH_FAILED(acct.getName(),
-                            namePassedIn(authCtxt), "invalid password");
-                }
-            } else if (PasswordUtil.SSHA.isSSHA(encodedPassword)) {
-                    if (PasswordUtil.SSHA.verifySSHA(encodedPassword, password)) {
-                        return; // good password, RETURN
-                    }  else {
+            if (isEncodedPassword(encodedPassword)) {
+                if (isValidEncodedPassword(encodedPassword, password)) {
+                    return;
+                } else {
+                    acct.refreshUserCredentials();
+                    String refreshedPassword = acct.getAttr(Provisioning.A_userPassword);
+                    if (!isEncodedPassword(refreshedPassword)) {
+                        doAuth(prov, domain, acct, password, authCtxt);
```

Comment:
This line bugs me.  Should we be returning here (e.g. ```return doAuth(...)```) or do we really intend to fall back through the code that follows after we make it through the subsequent call to doauth()?

Reply:
The old code did that very thing only throwing an exception in the case of a failure.  I was merely trying to maintain it's basic flow.

Reply:
The thought crossed my mind to add a 'return doAuth' there but was unsure if that was 'proper form' for Java considering it's a void return function.